### DISCUSSION TITLE
How to use ClientStateChange?

### DISCUSSION THREAD

**Discussion by bigfooted** [Kitware/trame#259] [Q&A] [answered]
What does ClientStateChange really do?
In one of the examples (VTPviewer), it is used like this:
`trame.ClientStateChange(value="mesh", change=ctrl.view_reset_camera)`

I would guess that when there is a change to state.mesh, then ctrl.view_reset_camera is called.
But when I just add this line to e.g. the cone tutorial, where the state.resolution is changed by a slider, and test it like this:
`trame.ClientStateChange(value="resolution", change=print("resolution has changed"))`
It is not being called (only once at initialization, not during resolution change). I could also remove this line from the example and did not notice any changes.
I guess I have to do something else as well to make this work properly, but couldn't figure out what.

**Comment by psavery** [comment, marked as answer]
About your example: the `change=` argument should be set to a function, but the `print()` call executes immediately and returns `None`. So, you need to change it to be something like this: `change=lambda: print("resolution has changed")` (if you get an error, you may also need to capture arguments, via `change=lambda *args, **kwargs: print("resolution has changed")`).

@jourdain can provide more details about `ClientStateChange`, but I believe it allows the client side to trigger the function call when the state changes on the client.

**Reply by jourdain** [reply]
Such trigger happen on the client side (which can call server method like any click=) but generally it should be used when you want to skip network call and keep the logic on the client side for speed (or with client only state variable).
